doc/man/mp4box.1
.TH "MP4Box" 1 "June 2005" "MP4Box" "GPAC"
.
.SH NAME
.LP 
MP4Box \- MPEG-4 Systems Toolbox
.SH "SYNOPSIS"
.LP 
.B MP4Box
.RI [options] \ file \ [options]
.br
.
.SH "DESCRIPTION"
.LP 
MP4Box is a multi-purpose command line tool to create and edit MPEG-4 Systems presentations and manipulate ISO-media files (MP4, 3GP, MOV).
.LP
MP4Box supports file conversion from various raw formats and IsoMedia/AVI/MPEG-PS/OGG containers, file hinting for RTP streaming for QuickTime compatible streaming servers, file interleaving, file fragmentation and track extraction.
.LP
MP4Box also provides dump tools used to inspect file layout, RTP hint tracks, SDP information, scene composition. It may also be used to convert to and from BT/XMT-A/VRML/X3D.
.LP
MP4Box also features MPEG-4 Systems encoders and decoders for BIFS and OD tools.
.LP
MP4Box doesn't expect any particular order in options at prompt.
.
.SH GETTING HELP WITH MP4Box
.TP
.B \-h [type]
prints help screen. "type" can be one of "general", "hint", "import", "encode", "extract", "dump", "swf", "meta", "crypt", "format", "rtp", "live" for specific help.
.TP
.B \-version
prints version of MP4Box.
.TP
.B \-noprog
disable progress reports.
.TP
.B \-quiet
quiet mode (no log info).
.TP
.B \-v
verbose mode (more log info).
.TP
.B \-strict-error
exits after the first error is reported.
.TP
.B \-nodes
lists supported MPEG-4 Systems nodes in MP4Box current build.
.TP
.B \-node NodeName
prints the node syntax. Each field is listed as
.br
.I eventType fieldType fieldName defaultValue
.br
and field quantization information and quantization bounds if any.
.
.TP
.B \-xnodes
lists supported X3D nodes in MP4Box current build.
.TP
.B \-xnode NodeName
prints the X3D node syntax. Each field is listed as
.
.TP
.B \-snodes
lists supported SVG/LASeR nodes in MP4Box current build.
.
.SH GENERAL OPTIONS
.P
.TP
.B \-inter duration
interleaves media data in chunks of the maximum specified duration (expressed in milliseconds) and prepare file for HTTP/FTP streaming. By default MP4Box always stores with 0.5 second interleaving. An interleaving value of '0' stores without interleaving (meta-data then track after track). 
.br
.B \-old-inter duration
same as -inter but without drift adjustment
.br
.B NOTE: Track Edit Lists are ignored when performing interleaving.
.TP
.B \-flat
stores file with all media data first, non interleaved. If used when creating a new file, no temporary file is created (faster storage).
.TP
.B \-frag duration
fragments file using ISO-Media movie fragments. Tracks will be fragmented so that no track run exceeds the specified duration (expressed in milliseconds). Disables interleaving.
.TP
.B -out\ \<filename>
specifies output file name. 
.TP
.B -tmp\ \<dirname>
specifies driectory for temporary storage. If not set, temporary storage is OS-dependent.
.TP
.B -ffspace size
inserts free space before moof in fragmented files.
.TP
.B \-nosys
removes all MPEG-4 Systems information from the file but keeps an empty IOD for MPEG-4 Profile@Level indications. alias: -no-sys
.TP
.B \-no-iod
removes InitialObjectDescriptor from file.
.TP
.B \-isma
rewrites the file as an ISMA 1.0 Audio/Video file (all systems info rewritten) with proper clock references.
.TP
.B \-ismax
rewrites the file as an ISMA 1.0 Audio/Video file (all systems info rewritten) with ISMA clock references (streams not explicetly synchronized).
.br 
.B NOTE: Seeking may be broken with some players (in ISMA, audio and video streams do not depend on main scene clock).
.TP
.B \-3gp
rewrites the file as a 3GP file (no more MPEG-4 Systems specific info). This option is turned on by default when input file extension is .3gp or .3g2.
.TP
.B \-ipod
rewrites the file for iPod.
.TP
.B \-brand ABCD[:v]
sets major brand of file, with optional version.
.TP
.B \-ab ABCD
adds given brand to file's alternate brand list.
.TP
.B \-rb ABCD
removes given brand to file's alternate brand list.
.TP
.B -cprt string
adds string copyright notice to file.
.TP
.B \-chap chap_file
adds chapter information contained in chap_file to movie. For more details on chapter file syntax, cf http://gpac.sourceforge.net/auth_mp4box.php.
.TP
.B -rem trackID
removes given track from file - can be specified multiple times.
.TP
.B -enable trackID
enables given track.
.TP
.B -disable trackID
disables given track.
.TP
.B -new
forces creation of a new destination file.
.TP
.B -lang [tkID=]LAN
sets the language of all tracks or the given track. LAN is the ISO 639-2 3 character code.
.TP
.B -delay [tkID=]delay_ms
sets the delay in milliseconds of all tracks or the given track.
.TP
.B -par tkID=PAR
sets visual track pixel aspect ratio (PAR=Num:Den or "none").
.TP
.B -name tkID=name
sets track handler name.
.TP
.B -itags tag1[:tag2]
sets iTunes tags to file - more info: MP4Box -tag-list.
.TP
.B \-split time_in_seconds
splits in files of desired maximum duration. This will remove all MPEG-4 Systems media.
.TP
.B \-split-size size_in_kilobytes
splits in files of desired maximum size. This will remove all MPEG-4 Systems media. Alias: -splits
.TP
.B \-split-chunk start:end
extracts a new file from specified start to end times (in seconds). This will remove all MPEG-4 Systems media. Alias: -splitx
.TP
.B \-splitz start:end
same as -split-chunk option, but adjust end time to be juste before the last rap found in the chunk.
.TP
.B \-split-rap start:end
splits in files begining at each RAP. This will remove all MPEG-4 Systems media. Alias: -splitr
.TP
.B \-group-add fmt
creates a new grouping information in the file. Format is a colon-separated list of following options:
.br
refTrack=ID: ID of the track used as a group reference. If not set, the track will belong to the same group as the previous trackID specified. If 0 or no previous track specified, a new alternate group will be created.
.br
switchID=ID: ID of the switch group to create. If 0, a new ID will be computed for you. If <0, disables SwitchGroup.
.br
criteria=string: list of space-separated 4CCs.
.br
trackID=ID: ID of the track to add to this group.
.br
*WARNING* Options modify state as they are parsed: trackID=1:criteria=lang:trackID=2 is different from: criteria=lang:trackID=1:trackID=2. 
.B \-group-rem-track ID
removes track from its group.
.TP
.B \-group-rem ID
removes the track's group.
.TP
.B \-group-clean
removes all group information from all tracks.
.TP
.B \-ref id:XXXX:refID
adds a reference of type 4CC from track ID to track refID.
.TP
.B \-dash DUR
enables DASH-ing of the file with a segment duration of DUR.
.TP
.B \-rap
segments begin with random access points.
.TP
.B \-frags-per-sidx N
sets the number of segments to be written in each SIDX box.
.TP
.B \-segment-name name
sets the segment name for generated segments.
.TP
.B \-segment-ext name
sets the segment extension name. Default is m4s.
.TP
.B \-url-template
uses UrlTemplate instead of explicit sources in segments.
.TP
.B \-daisy-chain
Uses daisy-chain SIDX instead of hierarchical. Ignored if frags/sidx is 0.
.TP
.B \-dash-ctx FILE
Stores/restore DASH timing from FILE.
.TP
.B \-dash-ts-prog N
Program_number to be considered in case of an MPTS input file.

.B When input file is an ISO-Media file (QT, MP4, 3GP), if no output is specified THE INPUT FILE IS OVERWRITTEN.
.
.
.SH HINTING OPTIONS
.TP
.B \-hint
hint the file for RTP\/RTSP sessions. Payload type is automatically detected and configured unless forced through one of MPEG-4 Generic RTP payload.
.TP
.B \-mtu size
specifies Maximum Transmission Unit size in bytes (eg maximum RTP packet size). Default size is 1500 bytes (Ethernet MTU). This must be choosen carefully: specifying too large packets will result in undesired packet fragmentation at UDP layer while specifying too small packets will result in RTP header overhead.
.TP
.B \-copy
forces hinted data to be copied to the hint track instead of simply referenced. This speeds up RTP packet construction at the server side but results in much bigger files.
.TP
.B \-tight
performs sample-based interleaving of media tracks and hint tracks. This should reduce disk seeks at server side (depending on server implementation) but results in a bigger file.
.TP
.B \-multi [maxptime]
enables Access Units concatenation in RTP packets if possible. maxptime is optional and specifies the maximum packet duration in milliseconds (default 100).
.TP
.B \-rate ck_rate
specifies the rtp rate in Hz when no default rate for payload. Default value is 90000 (MPEG rtp rates).
.TP
.B \-mpeg4
forces usage of the MPEG-4 generic payload whenever possible. Media tracks without a mapping to MPEG-4 Systems cannot use this.
.TP
.B \-latm
forces usage of the LATM payload for AAC audio.
.TP
.B \-static
enables usage of static RTP payload IDs for streams with official payload IDs. By default MP4Box always uses dynamic payload IDs for maximum interoperability, some players having troubles with static ones.
.
.P
.B MPEG-4 Generic Payload Options (Experts only)
.IP
.B \-ocr
forces all media tracks in the file to be served synchronized. This is needed because most streaming servers don't support desynchronized tracks in a single file. Be extremelly carefull when designing MPEG-4 interactive presentations for streaming since you will have to take care of the streaming server capabilities... MP4Box generates warnings when the file timeline can be ambiguously interpreted by the server.
.IP
.B \-rap
signals Access Units random access flag in RTP packets. This is usually only needed for streaming of MPEG-4 Systems streams.
.IP
.B \-ts
signals Access Units Time Stamps (CTS and DTS) in RTP packets.
.IP
.B \-size
signals Access Units size in RTP packets.
.IP
.B \-idx
signals Access Units indexes (sequence numbers) in RTP packets.
.IP
.B \-multi
enables Access Units concatenation in RTP packets (-ts, -size and -idx are selected if needed).
.IP
.B \-iod
prevents system tracks embedding in IOD (ISMA-like IOD) when generating in SDP. MP4Box automatically detects ambiguous (ISMA/non-ISMA) files but nobody's perfect. This shouldn't be used with -isma option.
.
.TP
.B \-add-sdp string
adds string to movie SDP or track SDP (tkID:string, where tkID is the OD of the hint track or its media track). This takes care of SDP line reordering, but not of SDP content validity.
.TP
.B \-unhint
removes all hint tracks and other hinting info from the file.
.
.SH IMPORT OPTIONS
.TP
.B \-add <src_file>
adds all src_file tracks to input file, creating it if not exisiting. Up to 20 cumulated -add operations can be used. Supported syntaxes are:
.br
file#video: imports first video track from src_file.
.br
file#audio: imports first audio track from src_file.
.br
file#trackID=ID or file#ID: imports given trackfrom src_file. To get a listing of tracks in input file, use -info [ID]
.br
[;lang=LAN]: specifies language of imported media.
.br
[;delay=delay_ms]: specifies initial delay in milliseconds of imported media.
.TP
.B \-cat <src_file>
concatenates all src_file tracks to input file, creating it if not exisiting. Media samples are added at the end of existing compatible tracks. If no compatible track is found for a media it is created. Up to 20 cumulated -cat operations can be used. Syntax is the same as -add.
.TP
.B \-force-cat
skips media configuration check when concatenating file.
.TP
.B \-keepsys
by default all MPEG-4 systems media are removed with -add and -cat. This option will avoid removing them from final file.
.TP
.B \-keep-all
keeps all existing tracks when add file.
.TP
.B \-dref
keeps media data in original file an only imports meta-data (frame timing, size and random access). 
.br
.TP
.B NOTE
Data referencing may fail with some AVI because it requires the framed data (eg an MP4 sample) to be continuous in the original file, which is not always the case depending on the original interleaving.
.TP
.B \-no-drop
forces constant FPS when importing AVI video. By default non coded frames (n-vop) are removed at import time, resulting in a variable frame-rate media.
.TP
.B \-packed
for CMP/M4V (raw MPEG-4 Visual), forces packed-bitstream mode (removes all n-vops and import at constant frame rate).
.TP
.B \-sbr
imports AAC as AAC-SBR, with backward compatible signaling (non AAC-SBR decoders should be able to play it).
.TP
.B \-sbrx
imports AAC as AAC-SBR, with non-backward compatible signaling (non AAC-SBR decoders will not be able to play it).
.TP
.B \-ovsbr
imports AAC as AAC-SBR with oversample SBR.
.TP
.B \-ps
imports AAC as AAC-PS, with backward compatible signaling of AAC-PS.
.TP
.B \-psx
imports AAC as AAC-PS, with non-backward compatible signaling (non AAC-PS decoders will not be able to play it).
.TP 
.B \-fps FrameRate
overrides the input video frame rate or specifies it for SUB subtitles.
.TP 
.B \-mpeg4
forces using MPEG-4 sample descriptions rather than 3GP ones (3GP2 QCELP/EVRC/SMV audio only).
.TP 
.B \-agg N
aggregates N audio frames in 1 sample (3GP media only). Maximum possible value is 15, and default value is 1 (no aggregation).
.
.SH ENCODING OPTIONS
.TP
.B \-mp4
specifies input file is for encoding. Supported inputs are BT/XMT-A/WRL/SWF files. Output file name is by default the input file name without extensions plus ".mp4" extension. 
.TP
.B \-def
encodes DEF'ed nodes and routes with their textual names.
.TP
.B \-log
generates BIFS encoder log file.
.TP
.B \-ms file
specifies file for track importing - by default FILE.mp4 is used when encoding FILE.bt (in-place rewrite). This option is only needed if you don't provide
a proper muxInfo per stream in the BT/XMT file.
.TP
.B \-sync time
forces BIFS random access point generation every time milliseconds. Cannot be used with -shadow. WARNING: this may result in weird behavior of your presentation since a BIFS random access point restarts all media currently running in the scene (EXPERTS ONLY - DANGEROUS). Cannot be used with -shadow.
.TP
.B \-shadow time
forces BIFS shadow random access points generation every time milliseconds. Shadow samples are random access points that can be used instead of non random access points when seeking. WARNING: this may be not supported by some players (EXPERTS ONLY - DANGEROUS). Cannot be used with -sync.
.TP
.B \-ctx-in file
specifies initial context (MP4/BT/XMTA) for chunk processing. Input file must then be a command-only file: no IOD, and no implicit commands (commands without 'AT').
.TP
.B \-ctx-out file
specifies output file of updated context (MP4/BT/XMTA) in chunk processing mode. This is optional, chunk processing doesn't need to store the final context.
.TP
.B \-resolution res
LASeR resolution factor (-8 to 7, default 0). All coords are multiplied by 2^res before truncation.
.TP
.B \-coord-bits bits
bits used for encoding truncated coordinates in LASeR. (0 to 31, default 12)
.TP
.B \-scale-bits bits
bits used for encoding truncated scales in LASeR. (0 to 4, default 0)
.TP
.B \-auto-quant res
Use automatic LASeR quantification. resolution is given as if using -resolution but coord-bits and scale-bits are infered.
.
.SH ISMACRYPT OPTIONS
.TP
.B \-crypt drm_file
crypts a specific track using ISMA AES CTR 128. 
.TP
.B \-decrypt [drm_file]
decrypts a specific track using ISMA AES CTR 128. drm_file can be omitted if keys are in file.
.TP
.B \-set-kms [tkID=]kms_uri
changes KMS location for all tracks or a given one if tkID is specified.
.TP
.B DRM file syntax for GPAC ISMACryp
File is XML and shall start with xml header. File root is an "ISMACryp" element. File is a list of "ISMACrypTrack" elements. 
.br
ISMACrypTrack attributes:
.br
TrackID: ID of track to en/decrypt.
.br
key:  AES-128 key formatted (hex string 0x +32 chars.
.br
salt: CTR IV salt key (64 bits) (hex string 0x +16 chars.
.br
Encryption only attributes
.br
Scheme_URI: URI of scheme used.
.br
KMS_URI: URI of key management system - \'self\' writes key and salt in the file.
.br
selectiveType selective encryption type - understood values are "None":              all samples encrypted (default), "RAP":  only encrypts random access units, "Non-RAP":  only encrypts non-random access units, "Rand": random selection is performed", "X": Encrypts every first sample out of X, "RandX": Encrypts one random sample out of X.
.br
ipmpType: IPMP Signaling Type: None, IPMP, IPMPX.
.br 
ipmpDescriptorID: IPMP_Descriptor ID to use if IPMP(X) is used. If not set MP4Box will generate one for you.
.

.SH EXTRACTING OPTIONS
.TP
.B \-raw TrackID
extracts given track in native format when supported.
.TP
.B \-raws TrackID
extract each track sample to a file. Note: 'TrackID:N' extracts Nth sample of the track.
.TP
.B \-nhnt TrackID
extracts given track in NHNT format. All track types except ObjectDescriptors tracks can be exported.
.TP
.B \-nhml TrackID
extracts track in nhml format (XML nhnt). All track types except ObjectDescriptors tracks can be exported.
.TP
.B \-single TrackID
extracts given track to a new mp4 file with a single track.
.TP
.B \-avi TrackID
extracts visual track to an avi file.
.TP
.B \-qcp TrackID
same as '-raw' but defaults to QCP file format for AVRC and SMV audio codecs.
.TP
.B \-aviraw TK
extracts AVI track to its raw format. TK can be one of "video", "audio" or "audioN" for multi-track avi files (cf '-info'). 
.TP
.B \-saf
remux file to SAF multiplex. 
.TP
.B \-dvbhdemux
demux DVB-H file into IP Datagrams. 
.TP
.B \-diod
extracts file IOD in raw format when supported. 
.
.SH DUMP OPTIONS
.TP
.B \-info [TrackID]
prints movie and tracks information. If TrackID specified, dumps only extended track info. If input file is not an IsoMedia file, lists known tracks for import.
.TP
.B \-bt
dumps complete scene in a BT file. This will remove unknown MPEG4 nodes.
.TP
.B \-xmt
dumps complete scene in an XMT-A file. This will remove unknown MPEG4 nodes.
.TP
.B \-wrl
dumps complete scene in an VRML97 WRL file. This will remove unknown VRML97 nodes.
.TP
.B \-x3d
dumps complete scene in an X3D XML file. This will remove unknown X3D nodes.
.TP
.B \-x3dv
dumps complete scene in an X3D Text (VRML) file. This will remove unknown X3D nodes.
.TP
.B \-lsr
dumps complete scene in a LASeR+XML file. 
.TP
.B \-diso
creates XML image of the file atoms.
.TP
.B \-drtp
creates XML image of all hint tracks samples of a hinted mp4 file.
.TP
.B \-dts
prints sample timing to text output.
.TP
.B \-dcr
creates XML image of all ISMACryp tracks samples of an mp4 file.
.TP
.B \-sdp
creates SDP file associated with a hinted mp4 file.
.TP
.B \-dump-cover
Extracts cover art if any.
.TP
.B \-dump-chap
Extracts chapter list to file if any.
.TP
.B \-ttxt
converts input subtitle to GPAC TTXT format.
.TP
.B \-ttxt trackID
exports given text track to GPAC TTXT format.
.TP
.B \-srt
converts input subtitle to SRT format.
.TP
.B \-srt trackID
exports given text track to SRT subtitle format.
.TP
.B \-std
dumps to stdout rather than file.
.TP
.B \-stat
generates statistics report on node/field usage for the whole presentation.
.TP
.B \-stats
generates statistics report on node/field usage per BIFS Access Unit.
.TP
.B \-statx
generates statistics report on node/field usage in the scene graph after each BIFS Access Unit.
.TP
.B \-hash
generates SHA-1 Hash of the input file.
.
.SH Meta OPTIONS
.
.LP
MP4Box supports the Meta addition to IsoMedia file format, used to make the file a generic data container (timed or untimed).
.TP
.B \-set-meta args
sets meta object type. Syntax is "ABCD[;tk=N] where:
.br
ABCD is the four character code of the type, or O/NULL to remove the meta object
.br
tk indicates whether the meta is at the root level (no "tk"), at the moov level (tk=0) or at the track level (tk=trackID).
.TP 
.B \-add-item args
adds resource to the given meta. args is the item file path followed by ';' separated options:
.br
tk=N: same as above
.br
name=item_name: specifies the item name, otherwise file name is used
.br
mime=mimeType: specifies the item mime type, otherwise application/octet-stream is used
.br
encoding=enctype: specifies the content encoding type
Note that file path set to "this" or "self" will identify the item added as the whole IsoMedia file
.TP
.B \-rem-item args
removes item from given meta. Syntax is itemID[;tk=ID].
.TP
.B \-set-primary args
sets given item as the primary one of the given meta. Syntax is itemID[;tk=ID].
.TP
.B \-set-xml args
sets XML to the given meta. Syntax is xml_file_path[;tk=ID][;binary].
.TP
.B \-rem-xml [tk=ID]
removes XML data from the given meta.
.TP
.B \-dump-xml xml_out_file_name[;tk=ID]
dumps the XML data of the given meta to a file.
.TP
.B \-dump-item itemID[;tk=ID][;path=fileName]
dumps the given item of the the given meta to a file. By default the item name is used as the output file name.
.TP
.B \-package
packages input XML file into an ISO container. all media referenced except hyperlinks are added to file.
.TP
.B \-mgt
packages input XML file into an MPEG-U widget in ISO container. All files contained in the current folder are added to the widget package.
.
.SH RTP STREAMER OPTIONS
.
.LP
MP4Box can stream ISO files to RTP. The streamer currently doesn't support data carrouselling and will therefore not handle BIFS and OD streams properly.      
.TP
.B \-rtp
enables rtp file streamer.
.TP
.B \-noloop
disables looping when streaming.
.TP
.B \-mpeg4
forces MPEG-4 ES Generic for all RTP streams.
.TP
.B \-dst=IP
IP destination (uni/multi-cast). Default: 127.0.0.1.
.TP
.B \-port=PORT
output port of the first stream. Default: 7000.
.TP
.B \-mtu=MTU
path MTU for RTP packets. Default is 1450 bytes.
.TP
.B \-ifce=IP
IP address of the physical interface to use. Default: NULL (ANY).
.TP
.B \-ttl=N
time to live for multicast packets. Default: 1.
.TP
.B \-sdp=FILE
file name of the generated SDP. Default is session.sdp.
.
.SH LIVE SCENE STREAMER OPTIONS
.
.LP
MP4Box can stream BIFS or DIMS content to RTP with carousel generation and updates. The streamer currently doesn't support audio/video/image streaming at the same time. All options supported by the file streamer are supported.
.TP
.B \-live
enables rtp live streamer.
.TP
.B \-dims
turns on DIMS mode for SVG input - default: off.
.TP
.B \-src=FILE
source of updates - default: null.
.TP
.B \-rap=TIME
duration in ms of base carousel - default: 0 (off). you can specify the RAP period of a single ESID (not in DIMS) with -rap=ESID=X:time.
.TP
.B Runtime Options
The following options can be used at prompt:
.br
q: quits
.br
u: inputs some commands to be sent
.br

Changelog
27/06/2019: GPAC 0.8.0
 - General
    * Many security fixes (static compile and fuzzing through AFL, always ongoing).
    * Many bugs fixes
    * Added :ncl option in log levels to disable color logs
    * More tests and coverage
 - File Formats
    * Better support of QTFF / ProRes files
    * Support for AV1
       - import and export
       - source formats: OBUs (Section 5), IVF and AnnexB
       - AV1 in HEIF
    * Support for color (nclc, nclx and ICC profiles) in HEIF and ISOBMFF
    * Support for HDR (mdcv, clli) info in HEIF and ISOBMFF
    * Support for alpha in HEIF
    * Support for enforcing pasp presence even for 1:1 ratios
    * HEVC temporal sublayer split in MP4Box
    * Allow meta storage before mdat for meta-only files (heif and co)
    * Added option to keep AU delimiter in isobmff samples
    * Support for opus import
    * Support for pixi and ccst in HEIF
    * DolbyVision 'dvcC' and partial 'dvhe' boxes for HDR
    * Support for VP9 import and playback
    * Sample dependency in avc and hevc importers, and track thinner for non-refs images
    * Support for audio_roll signaling
    * New audio import mode to control AudioSampleEntry creation (v0, v1 ISOBMFF, v1 QTFF)
    * xHE-AAC import with detection of sync samples
    * Added support for MPEG-H 3D audio boxes (no import yet)
    * Handle Vobsub empty SPU packets
    * Added auxv and pict support
 - Common Encryption
    * Fully compliant CENC supporting cenc, cens, cbc1 and cbcs
    * CENC for AV1
    * Improved DASH+CENC support, pssh in MPD
    * ForceClear mode for CENC to skip encryption without sample groups
    * Made senc in movie fragments always stored before truns
    * Added default values handling for cbcs and possibility to set protection system per track
    * Compatibility with OpenSSL 1.1.x
 - Streaming and Adaptive Streaming
    * Support for ATSC3.0 both US and Korean versions !
    * Support for for live splices (xlink period insertions) in DASH client
    * Automatic period continuity in DASH when no codec change between periods
    * Added DASH cue-base segmentation (XML based) and -dsap option to generate cue files from source
    * Support for BBA-0 and BOLA implementations
    * Write fragment defaults in trex even when not using them
    * Support for simple ssix for keyframe data byterange at the start of a segment
    * Moved segment template at AdaptationSet level if only one representation
    * Changed default bsmode in dasher if single input file
    * Added init-seg-ext option
    * Added -mvex-after-traks option to MP4Box when dashing for CMAF
    * Added segmentation option to insert a tfdt per traf
    * Added -closest mode for DASH segmentation
    * Added -bound option to use audio segmenting method for video
    * Renamed -dash-run-for to -run-for
    * Added '=' in dash templates
    * Improved bandwith estimation when using HTTP 1.1 chunk transfer
    * Add option to force moof base offsets
 - MP4Box
    * Added -catpl to concatenate from playlist in MP4Box
    * Added options to set movie timescale at import and dash time
    * Added mpd rip option and top-level box compressor in MP4Box
    * Made -dts skip timing check and added -dtsc for that
    * Made force-cat option more agressive
    * Support for MovieFragmentRandomAccess using -mfra option
    * Added -dtsx to dump timing without offset
    * Added -dnalc opt for nal CRC dump
    * Added chunk extraction up to time until end
    * ISOBMFF single track import now removes references by default
 - Decoders
    * Updated ffmpeg to 4.0.2
    * Moved to openHEVC 3.0 API
    * Added nvdec support (windows, linux) with reuse of decoder context for tiled VR
    * Added HEVC support to mediacodec on android
    * AV1 playback through ffmpeg
    * Opus playback through ffmpeg
 - 3D, VR and 360
    * Added vrhud for multiviewpoint 360
    * Added forced visibility mode of tiles in VR
    * Added tile visibility debug mode
    * Added forced stereo output for openhevc
    * Disable face nav if mouse grabbed
    * Added simple face tracking vr navigation based on udp commands
    * Added PSVR support
    * Added mouse move emulation at window border to force sphere rotation when inactive
    * Changed tile visibility algo to sample points in mesh
 - Players (Mobile and Desktop)
    * Added about extension
    * Added multiple audio objects in dynamic scene
    * Added addon splicing of main content
    * Added mosaic://v1:.:vN url support
    * Added gaze simulation through mouse and gaze-sphere visibility test
 - Subtitles
    * Allow * as argument of -srt|ttxt to dump all possible tracks (#925)
    * Improved support for WebVTT import
    * Improved support for WebVTT DASHing/fragmentation
 - Misc
    * OSX install now done through PKG and modify PATH env in/etc/paths
    * Added initial PMT version and disc marker to TS muxer
    * Moved dektec output to matrix API, added SDI clipping
    * Added temi periodic toggle and manual toggle in MP42TS

26/04/2017: GPAC 0.7.1
 - Full changelog at https://github.com/gpac/gpac/wiki/GPAC-0.7.0
 - Many security fixes (static compile and fuzzing through [AFL](http://lcamtuf.coredump.cx/afl/), always ongoing)
 - Colorized log.
 - Fix pkg-config Private.libs.
 - Changed default audio volume to 100% instead of 75%.
 - Expose more experimental options through the [documentation](http://htmlpreview.github.io/?https://github.com/gpac/gpac/blob/master/doc/configuration.html).
 - Improved GLES renderer on mobile platforms 
 - YUV422 and YUV444 8 and 10 bit support in GLES renderer 
 - Improvements on AVI dump.
 - SAT>IP modified RTSP client.
 - Added L-HEVC File Format support (SHVC/MV-HEVC tracks and HEVC Tile Tracks).
 - Added MPEG IFF (image File Format) support.
 - Range extension support for AVC and HEVC.
 - SHVC and MV-HEVC importers and playback: moved to final spec version (SHM6+).
 - Support of HLS with fragmented MP4 playback.
 - APIs: gf_mpd_() functions and new segmenter API.

applications/mp4box_android/src/com/enst/mp4box/mp4box.java
package com.enst.mp4box;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import com.enst.mp4box.R;

import android.app.Activity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.util.Log;
import android.view.View;



public class mp4box extends Activity {
	private mp4terminal myTerminal;
	
    /** Called when the activity is first created. */
    @Override
    
    
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        myTerminal = new mp4terminal();
        errors = loadAllLibraries();
        System.out.println( "hello world java" );
        final Button button = (Button) findViewById(R.id.OkButton);
        
        button.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
            	EditText oCommandLine = (EditText) findViewById(R.id.CommandLineEdit );
            	CharSequence sCommandLine = oCommandLine.getText();
            	myTerminal.run( sCommandLine.toString() );
            	//showKeyboard( true );
            }
        });
        
    }
    

    
    
    private static Map<String, Throwable> errors = null;
    
    synchronized static Map<String, Throwable> loadAllLibraries() 
    {
    	if( errors != null)
    		return errors;
        StringBuilder sb = new StringBuilder();
        final String[] toLoad = { "GLESv1_CM", "dl", "log",//$NON-NLS-3$ //$NON-NLS-2$ //$NON-NLS-1$
                             "jpeg", "javaenv", //$NON-NLS-1$ //$NON-NLS-2$ 
                             "mad", "editline", "ft2", //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
                             "js_osmo", "openjpeg", "png", "z", //$NON-NLS-1$ //$NON-NLS-2$//$NON-NLS-3$ //$NON-NLS-4$
                             "ffmpeg", "faad", "gpac", //$NON-NLS-1$//$NON-NLS-2$ //$NON-NLS-3$
                             "stdc++", "mp4box" }; // //$NON-NLS-1$ //$NON-NLS-2$
        HashMap<String, Throwable> exceptions = new HashMap<String, Throwable>();
    for (String s : toLoad) {
        try {
            String msg = "Loading library " + s + "...";//$NON-NLS-1$//$NON-NLS-2$
            sb.append(msg);
            //Log.i(LOG_LIB, msg);
            System.loadLibrary(s);
        } catch (UnsatisfiedLinkError e) {
            sb.append("Failed to load " + s + ", error=" + e.getLocalizedMessage() + " :: " //$NON-NLS-1$//$NON-NLS-2$//$NON-NLS-3$
                      + e.getClass().getSimpleName() + "\n"); //$NON-NLS-1$
            exceptions.put(s, e);
            //Log.e(LOG_LIB, "Failed to load library : " + s + " due to link error " + e.getLocalizedMessage(), e); //$NON-NLS-1$ //$NON-NLS-2$
        } catch (SecurityException e) {
            exceptions.put(s, e);
            //Log.e(LOG_LIB, "Failed to load library : " + s + " due to security error " + e.getLocalizedMessage(), e); //$NON-NLS-1$ //$NON-NLS-2$
        } catch (Throwable e) {
            exceptions.put(s, e);
            //Log.e(LOG_LIB, "Failed to load library : " + s + " due to Runtime error " + e.getLocalizedMessage(), e); //$NON-NLS-1$ //$NON-NLS-2$
        }
    }

    /*if (!exceptions.isEmpty()) {
        try {
            PrintStream out = new PrintStream(config.getGpacConfigDirectory() + "debug_libs.txt", "UTF-8"); //$NON-NLS-1$//$NON-NLS-2$
            out.println("$Revision: 2972 $"); //$NON-NLS-1$
            out.println(new Date());
            out.println("\n*** Configuration\n"); //$NON-NLS-1$
            out.println(config.getConfigAsText());
            sb.append("*** Libs listing: "); //$NON-NLS-1$
            sb.append(config.getGpacLibsDirectory());
            sb.append('\n');
            listing(sb, new File(config.getGpacLibsDirectory()), 2);
            sb.append("*** Modules listing: "); //$NON-NLS-1$
            sb.append(config.getGpacModulesDirectory());
            sb.append('\n');
            listing(sb, new File(config.getGpacModulesDirectory()), 2);
            sb.append("*** Fonts listing: \n"); //$NON-NLS-1$
            sb.append(config.getGpacFontDirectory());
            sb.append('\n');
            listing(sb, new File(config.getGpacFontDirectory()), 2);
            sb.append("*** Exceptions:\n"); //$NON-NLS-1$
            for (Map.Entry<String, Throwable> ex : exceptions.entrySet()) {
                sb.append(ex.getKey()).append(": ") //$NON-NLS-1$
                  .append(ex.getValue().getLocalizedMessage())
                  .append('(')
                  .append(ex.getValue().getClass())
                  .append(")\n"); //$NON-NLS-1$
            }
            out.println(sb.toString());
            out.flush();
            out.close();
        } catch (Exception e) {
            Log.e(LOG_LIB, "Failed to output debug info to debug file", e); //$NON-NLS-1$
        }
    }*/
    errors = Collections.unmodifiableMap(exceptions);
    return errors;
}
    /*public void showKeyboard(boolean showKeyboard) {
        if (keyboardIsVisible == showKeyboard == true)
            return;
        InputMethodManager mgr = ((InputMethodManager) getSystemService(INPUT_METHOD_SERVICE));
        this.keyboardIsVisible = showKeyboard;
        if (showKeyboard)
            mgr.showSoftInput(findViewById(R.id.CommandLineEdit ), 0);
        else
            mgr.hideSoftInputFromInputMethod(findViewById(R.id.CommandLineEdit ).getWindowToken(), 0);

    }*/
}